src/tracers.jl

```diff
@@ -12,147 +12,142 @@ The provided index set type `S` has to satisfy the following conditions:
 Subtypes of `AbstractSet{<:Integer}` are a natural choice, like `BitSet` or `Set{UInt64}`.
 """
 
+empty_sparse_vector(T) = T()
+empty_sparse_matrix(T) = T()
+
+sparse_vector(T, index) = T([index])
+sparse_vector(::Type{T}, index) where {T<:DuplicateVector} = T(index)
+sparse_vector(::Type{T}, index) where {T<:SortedVector} = T(index)
+
+#================#
+# Elementwise OR #
+#================#
+
+## We use ∨ to represent the elementwise OR
+# TODO REVIEW: \vee looks too much like the character 'v' in code
+# Gradient representations
+∨(a::G, b::G) where {G<:AbstractSet} = a ∪ b
+∨(a::G, b::G) where {G<:DuplicateVector} = G(vcat(a.data, b.data))
+∨(a::G, b::G) where {G<:SortedVector} = a ∪ b
+∨(a::G, b::G) where {G<:RecursiveSet} = a ∪ b
+
+# Hessian representations
+# TODO REVIEW: for now, we only support representing Hessians as a set of tuples
+∨(a::H, b::H) where {I<:Integer,H<:AbstractSet{Tuple{I,I}}} = a ∪ b
+
+## Outer product on gradients
+# Compute `out ∨ (𝟙[∇a] ∨ 𝟙[∇b]ᵀ)` in out
+function outer_product_or!(
+    out::H, a::G, b::G
```

Comment:
why in-place?

Reply:
I suggest the following:
1. I replace the inplace mutations with code matching the theory 1-to-1
2. We add meaningful Hessian benchmarks here or in a follow-up PR 
3. We improve performance in another follow-up PR

Reply:
I'm 100% in favor of that